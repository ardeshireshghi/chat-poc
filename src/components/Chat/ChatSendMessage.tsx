import React, { useCallback, useState } from 'react';
import styled from 'styled-components';
import { normalSpacing } from '../../styleguide/utils';
import TextInput from '../TextInput';
import Button from '../Button';

const ChatSendMessageForm = styled.form`
  display: flex;
  gap: ${normalSpacing};
  padding: ${normalSpacing};
  border-top: 1px solid ${(props) => props.theme.colors.lightGray};
`;

const SendButton = styled(Button)``;

interface ChatSendMessageProps {
  onMessageSent: (message: string) => void;
}
export const ChatSendMessage: React.FC<ChatSendMessageProps> = ({ onMessageSent }) => {
  const [message, setMessage] = useState('');

  const handleSubmit = useCallback(
    (event) => {
      event.preventDefault();
      if (message) {
        onMessageSent(message);
        setMessage('');
      }
    },
    [message, onMessageSent]
  );

  return (
    <ChatSendMessageForm onSubmit={handleSubmit}>
      <TextInput
        value={message}
        onChange={(e) => setMessage(e.target.value)}
        type="text"
        placeholder="Send a message..."
      ></TextInput>
      <SendButton>Send</SendButton>
    </ChatSendMessageForm>
  );
};
