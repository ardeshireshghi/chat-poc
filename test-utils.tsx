import React from 'react';
import { render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';
import { defaultTheme } from './src/styleguide/theme';

import { ChatDataProvider } from './src/context/chatContext';
import { IMessage } from './src/domain/models/message';

const renderWithThemeProvider = (ui: React.ReactElement, options = {}) => {
  const Wrapper: React.FC = ({ children }) => {
    return <ThemeProvider theme={defaultTheme}>{children}</ThemeProvider>;
  };
  return render(ui, { wrapper: Wrapper, ...options });
};

const renderWithAllProviders = (
  ui: React.ReactElement,
  { chatInitialMessages = [], ...options }: { chatInitialMessages?: IMessage[] } = {}
) => {
  const AllTheProviders: React.FC = ({ children }) => {
    return (
      <ThemeProvider theme={defaultTheme}>
        <ChatDataProvider value={{ messages: chatInitialMessages }}>{children}</ChatDataProvider>
      </ThemeProvider>
    );
  };
  return render(ui, { wrapper: AllTheProviders, ...options });
};

// re-export everything
export * from '@testing-library/react';

// override render method
export { renderWithAllProviders, renderWithThemeProvider };
