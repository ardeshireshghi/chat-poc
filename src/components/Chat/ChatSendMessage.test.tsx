import React from 'react';
import userEvent from '@testing-library/user-event';
import { renderWithThemeProvider } from '../../../test-utils';
import { ChatSendMessage } from './ChatSendMessage';

describe('ChatSendMessage', () => {
  it('renders correctly', () => {
    const messageSentMock = jest.fn();
    const { container } = renderWithThemeProvider(
      <ChatSendMessage onMessageSent={messageSentMock} />
    );

    expect(container).toMatchSnapshot();
  });

  it('calls message sent handler when chat message is sent', async () => {
    const messageSentMock = jest.fn();
    const { getByRole, getByPlaceholderText } = renderWithThemeProvider(
      <ChatSendMessage onMessageSent={messageSentMock} />
    );

    const messageInput = getByPlaceholderText(/Send a message/);
    const sendBtn = getByRole('button');

    userEvent.type(messageInput, 'Hello, World!');
    userEvent.click(sendBtn);

    expect(messageSentMock).toHaveBeenCalledWith('Hello, World!');
  });
});
