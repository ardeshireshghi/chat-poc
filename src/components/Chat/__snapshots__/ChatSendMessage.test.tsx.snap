// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatSendMessage renders correctly 1`] = `
<div>
  <form
    class="sc-dlfnbm ibzwYW"
  >
    <input
      class="sc-bdfBwQ bTUsFP"
      placeholder="Send a message..."
      type="text"
      value=""
    />
    <button
      class="sc-gsTCUz sc-hKgILt kkzXpY"
    >
      Send
    </button>
  </form>
</div>
`;
